src/gpu/generic/sycl/sycl_io_helper.hpp

```diff
@@ -184,6 +184,70 @@ inline void store_float_vec(data_type_t dt, ::sycl::vec<float, width> vec_f32,
 #undef CASE
 }
 
+template <::sycl::access_mode mode>
+struct memory_plain_t {
+    memory_plain_t(xpu::sycl::memory_arg_t<mode> mem, data_type_t type)
```

Comment:
While I understand the problem that the newly introduced `memory_plain_t` and `memory_tensor_t` classes are supposed to solve I think it may not be the best solution. There are already a few memory related abstractions in the library and there is already a related redundancy that we will need to remove so I think we should move in the opposite direction, instead of introducing new abstractions, we should generalize the ones we already have.

One of the options is the following:
* Get rid of `interop_memory_arg_t` by replacing it with `memory_arg_t`
* Extend `memory_arg_t` to contain `sycl_md_t` or just data type. We could introduce a state to `memory_arg_t` to be able to check what interfaces can be used prior to using them
* Introduce load/store interfaces to `memory_arg_t`
 
This way we will have just one memory argument class that can do it all.

Reply:
I do agree that there is redundancy in the amount of abstractions we have and I would like to move into direction of having just one for all use cases. However this and existing abstractions are solving different problems (the two I introduced simplify loading and storing of data, `md_t` store everything about layout, `interop_memory_arg_t` is used for interop) and are used in different places. 

I have so far not worked directly with `interop_memory_arg_t` so I am not aware of any deficiencies. As far as I can see it is mostly equivalent to existing `memory_arg_t`. But both of those are lower level abstractions - just abstracting away which type of memory do we have (buffer/usm). I see value in unifying those two, but that is going in a different direction than this PR. The abstractions I am introducing here are building on top of these. I think they make more sense in separate classes.

Eventually I would like to build more on the newly introduced abstractions so that they can be the only one that sycl kernels need to use.

In general I do not fully understand what you are suggesting, so I find what I am introducing here more useful. If you still believe your suggestions are better, maybe you could share the proposed code changes?

Reply:
It seems that we are talking about very similar objectives, the difference seems to be in the implementation details. In a nutshell, my suggestion was to add `sycl_md_t` to `memory_arg_t` along with the load/store interfaces. That would allow us to remove `sycl_md_t` from conf structures and let each kernel decide what memory descriptors it needs (e.g. forward convolution now has to contain memory descriptors for backward pass because there is only one conf structure for convolution).

I agree with you that keeping `memory_arg_t` as an abstraction for raw memory arguments for SYCL kernels is a good option too. Ultimately, it is up to you how you want to proceed.

If you want to proceed with your solution then I've got a couple of questions about the names and semantics of the classes.
> Eventually I would like to build more on the newly introduced abstractions so that they can be the only one that sycl kernels need to use.

* What do you mean by ` so that they can be the only one that sycl kernels need to use`? Do you mean that the new classes will replace `memory_arg_t` in the SYCL kernels (e.g. `xpu::sycl::in_memory_arg_t src_` -> `generic::sycl::memory_tensor_t src_`) or that the new classes will only be used as helpers during the kernel execution?
* The difference between `memory_tensor_t` and `memory_plain_t` classes seems to be that the latter doesn't contain a memory descriptor, instead it contains a data type. As far as I understand it's meant to be used for either scalars or 1D tensors. Is it correct?

Reply:
Thanks for the explanation, I think I get what you are suggesting now. Your approach would lead to simpler code, but potentially less performant one. 

Currently and in this PR multiple memory objects can share the same `md_t`. Since multiple `memory_tensor_t`s using the same `md_t` are constructed in kernel, the compiler can optimize that and only store the reused `md_t` once. 

If the same data is moved to `memory_arg_t` it will be copied into the kernel from the host. AFAIK due to the two pass architecture the compiler does not have visibility into how data is transferred into the kernel, so it can not optimize that. That would lead into some `md_t`s being duplicated in device registers, increasing the register pressure, which can lead to lower occupancy and lower performance. Also on Intel GPUs there is quite a tight limit on how much data can be passed to the kernel as arguments.

So Ideally we need to construct `memory_tensor_t` in kernel from `md_t` and `memory_arg_t`. In some kernels we could go even further to deduplicate data between different `md_t`s passed to the kernel, but that is not a part of this PR.

Reply:
> Currently and in this PR multiple memory objects can share the same md_t. Since multiple memory_tensor_ts using the same md_t are constructed in kernel, the compiler can optimize that and only store the reused md_t once.

I'm not a compiler expert but this optimization seems to be too difficult for a compiler to implement in practice, I wouldn't count on it for sure. But maybe I'm missing something.

> If the same data is moved to memory_arg_t it will be copied into the kernel from the host. 

Yes, but we copy it as part of conf structure anyway so if we remove `md_t` from conf structure as I previously mentioned then there won't be any additional overhead.

> They are meant to be used for data of any dimensionality where the information about dimensionality is not needed in the kernel. 

So, `memory_plain_t` is basically just a raw pointer with the load/store interface. I'm confused because I don't understand why this code is better: 
```cpp
memory_tensor_t diff_dst_mem(diff_dst_, conf_.diff_dst_md);
float dd = diff_dst_mem.load(diff_dst_off);
```
than this:
```cpp
// Assuming we changed `load_float_value` interface to take `memory_arg_t` and `md_t`
float dd = load_float_value(diff_dst_, conf_.diff_dst_md, diff_dst_off);
```

I'm just trying to make sure that we don't increase maintenance cost without getting anything in exchange.

Reply:
> But maybe I'm missing something.

As far as I understand the compiler will mostly inline everything, replace structs with individual members (this part AFAIK can not be done for kernel arguments), than it is easy to figure out the same numbers are used in calculations multiple times.

> if we remove md_t from conf structure as I previously mentioned then there won't be any additional overhead.

`md_t` is still copied from host to kernel. Now each one is copied once. If we reuse it in multiple objects and all of these objects are copied to the kernel, it will be copied and be stored in registers multiple times.

> I'm confused because I don't understand why this code is better than this

I prefer is for two reasons: First in some kernels one `memory_tensor_t` is used multiple times. In that case only small part of the code need s to be duplicated - `diff_dst_mem.load(diff_dst_off);`. Second, the type and the pointer are always used together (in some cases md_t as well). So it makes sense for them to be in the same object to make code at the point where data is loaded easier to understand and to reduce chance of errors where wrong combination of them is used together.

Reply:
Ok, I don't have strong arguments against your preferences so I think we shell move on and merge the changes. Thanks for the discussion!